### PR TITLE
 Make linter to enforce our preferred types order in type switch

### LINKED ISSUE #527: Make linter to enforce our preferred types order in type switch
In many places in our code, we have type switches that check all or some of BSON types:
```go
switch v := v.(type) {
case *types.Document:
  // ...
case *types.Array:
  // ...
case float64:
  // ...
```

We have a preferred types order that is a bit strange but follows the order in the BSON spec: https://pkg.go.dev/github.com/FerretDB/FerretDB/internal/types#hdr-Mapping

Sometimes, we forget about that order. We should make a linter that enforces it.

Can it be done with https://github.com/quasilyte/go-ruleguard?

### PR DESCRIPTION
**Pull request by fenogentov**
Closes #527.

### CODE REVIEW

**Review by AlekSi — CHANGES_REQUESTED** [review]
Let's move the tool into tools/check-switch.go, hook it into tools/tools.go, tools/go.mod

File: go.mod [hunk 1]
```diff
@@ -37,5 +37,8 @@ require (
 	go.uber.org/atomic v1.7.0 // indirect
 	go.uber.org/multierr v1.6.0 // indirect
 	golang.org/x/text v0.3.7 // indirect
+	golang.org/x/tools v0.1.10
 	google.golang.org/protobuf v1.26.0 // indirect
 )
+
+require golang.org/x/mod v0.6.0-dev.0.20220106191415-9b9b3d81d5e3 // indirect
```
Comment (AlekSi) [inline go.mod:44]: Both those changes should not be there

**Comment by seeforschauer** [comment]
@fenogentov would you please fix linters?

**Comment by fenogentov** [comment]
> @fenogentov would you please fix linters?

this is an example of how the new linter works

File: tools/tools.go [hunk 2]
```diff
@@ -36,6 +36,8 @@ import (
 
 //go:generate go run check.go
 
+//go:generate go install check-switch.go
```
Comment (AlekSi) [inline tools/tools.go:39]: Let build it into `../bin/` like we do with other tools

File: Taskfile.yml [hunk 3]
```diff
@@ -170,6 +170,7 @@ tasks:
       - lint-go-consistent-integration
     cmds:
       - bin/golangci-lint run --config=.golangci-new.yml --new-from-rev=main
+#      - go vet -vettool=$(which check-switch) ./...
```
Comment (AlekSi) [inline Taskfile.yml:173]: I think `../bin/check-switch` should work

File: tools/check-switch/check-switch_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	"golang.org/x/tools/go/analysis/analysistest"
+)
+
+func TestCheckSwitch(t *testing.T) {
+	path, err := os.Getwd()
+	if err != nil {
+		log.Panicln(err)
+	}
+	testdata := filepath.Join(path, "testdata")
+	analysistest.Run(t, testdata, Analyzer, "a")
```
Comment (AlekSi) [inline tools/check-switch/check-switch_test.go:32]: Shouldn't we check results?

File: Taskfile.yml [hunk 5]
```diff
@@ -170,6 +170,7 @@ tasks:
       - lint-go-consistent-integration
     cmds:
       - bin/golangci-lint run --config=.golangci-new.yml --new-from-rev=main
+#      - go vet -vettool=$(which ./bin/check-switch) ./...
```
Comment (AlekSi) [inline Taskfile.yml:173]: We should run the linter

File: tools/check-switch/testdata/src/a/a.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
+package a
```
Comment (AlekSi) [inline tools/check-switch/testdata/src/a/a.go:1]: Can it be just in `testdata/a.go`

File: Taskfile.yml (hunk 5 above)
Comment (AlekSi) [reply Taskfile.yml:173]: Let's start with better error messages :) Then please send PRs that fix those problems

File: Taskfile.yml [hunk 7]
```diff
@@ -191,6 +191,7 @@ tasks:
       - lint-go-consistent-integration
     cmds:
       - bin/golangci-lint run --config=.golangci-new.yml --new-from-rev=main
+      - go vet -vettool=$(which ./bin/check-switch) ./...
```
Comment (AlekSi) [inline Taskfile.yml:194]:
```suggestion
      - go vet -vettool=./bin/check-switch ./...
```

There should not be a need to use `which` if we provide a path

File: tools/tools.go [hunk 8]
```diff
@@ -36,6 +36,7 @@ import (
 
 //go:generate go run check.go
 
+//go:generate go build -v -o ../bin/ check-switch/check-switch.go
```
Comment (AlekSi) [inline tools/tools.go:39]:
```suggestion
//go:generate go build -v -o ../bin/ ./check-switch
```

To use package path as in other lines

File: tools/check-switch/testdata/src/a/a.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+		fmt.Println(v)
+	case bool:
+		fmt.Println(v)
+	case time.Time:
+		fmt.Println(v)
+	case types.NullType:
+		fmt.Println(v)
+	case types.Regex:
+		fmt.Println(v)
+	case int32:
```
Comment (AlekSi) [inline tools/check-switch/testdata/src/a/a.go:32]:
```suggestion
	case int32, int64:
```

How that case will be handled? Let's test it

File: tools/check-switch/check-switch.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+							case *ast.SelectorExpr:
+								name = fmt.Sprintf("%s.%s", cs.X, cs.Sel.Name)
+
+							case *ast.Ident:
+								name = cs.Name
+							}
+
+							subidxSl, ok := orderTypes[name]
+							if ok && (subidxSl < subidx) {
+								pass.Reportf(n.Pos(), "non-observance of the preferred order of types")
```
Comment (AlekSi) [inline tools/check-switch/check-switch.go:104]: Can we add types to the error message?

File: tools/check-switch/check-switch.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+	Name: "checkerswitch",
+	Doc:  "checking the preferred order of types in the switch",
+	Run:  run,
+}
+
+func main() {
+	singlechecker.Main(Analyzer)
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
```
Comment (AlekSi) [inline tools/check-switch/check-switch.go:52]: That function would benefit from comment explaining how it works.
And let's add some comments inside

File: tools/check-switch/check-switch.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+// orderTypes is preferred order of the types in the switch.
+var orderTypes = map[string]int{
+	"Document":  0,
+	"Array":     1,
+	"float64":   2,
+	"string":    3,
+	"Binary":    4,
+	"ObjectID":  5,
+	"bool":      6,
+	"time.Time": 7,
```
Comment (AlekSi) [inline tools/check-switch/check-switch.go:34]: Why only `time.Time` has a package qualifier? Why not, for example, `types.Timestamp` below?

File: internal/handlers/common/filter.go [hunk 13]
```diff
@@ -1025,14 +1034,14 @@ func filterFieldValueByTypeCode(fieldValue any, code typeCode) (bool, error) {
...
 	switch code {
-	case typeCodeObject:
-		if _, ok := fieldValue.(*types.Document); !ok {
-			return false, nil
-		}
 	case typeCodeArray:
 		if _, ok := fieldValue.(*types.Array); !ok {
 			return false, nil
 		}
+	case typeCodeObject:
```
Comment (AlekSi) [inline internal/handlers/common/filter.go:1041]: Object/document should go before array. I guess that change is unrelated and not caught by the linter?

**Review by AlekSi — CHANGES_REQUESTED** [review]

**Review by AlekSi — COMMENTED** [review]
@fenogentov How do you run this test? It does not work for me:

```
~/C/F/F/t/checkswitch (pr/fenogentov/654) -> go test

src/a/a.go:7:2: cannot find package "github.com/FerretDB/FerretDB/internal/types" in any of:
        /opt/homebrew/Cellar/go/1.18.3/libexec/src/github.com/FerretDB/FerretDB/internal/types (from $GOROOT)
        /Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/github.com/FerretDB/FerretDB/internal/types (from $GOPATH)
/Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/a/a.go:7:2: could not import github.com/FerretDB/FerretDB/internal/types (invalid package name: "")
/Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/a/a.go:45:7:        previous case
/Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/a/a.go:74:7:        previous case
/Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/a/a.go:105:7:       previous case
/Users/aleksi/Code/FerretDB/FerretDB/tools/checkswitch/testdata/src/a/a.go:136:7:       previous case
--- FAIL: TestCheckSwitch (0.19s)
    analysistest.go:296: error analyzing checkswitch@a: analysis skipped due to errors in package
FAIL
exit status 1
FAIL    github.com/FerretDB/FerretDB/tools/checkswitch  0.327s
```

File: tools/checkswitch/checkswitch.go [hunk 14]
```diff
@@ -24,61 +24,33 @@ import (
...
-	"timestampType":     12,
-	"typeCodeTimestamp": 12,
-
-	"int64":        13,
-	"int64Type":    13,
-	"typeCodeLong": 13,
-
-	"typeCodeNumber": 14,
-
-	"CString": 15,
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply tools/checkswitch/checkswitch.go]: @fenogentov ^ I think we should keep type codes in that map and group items

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/654#discussion_r914470334

File: tools/checkswitch/checkswitch.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply tools/checkswitch/checkswitch.go]: > group items

That was done

> keep type codes

That wasn't :) See affe64fefe55d8842c5b0d8f085ecdfb9e508fbb

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/654#discussion_r915493790

**Review by AlekSi — COMMENTED** [review]
Please resolve merge conflicts

**Review by AlekSi — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

**Pull request merged**